container-core/src/main/java/com/yahoo/container/core/documentapi/MessageBusDocumentAccessProvider.java

```diff
@@ -0,0 +1,55 @@
+package com.yahoo.container.core.documentapi;
+
+import com.google.inject.Inject;
+import com.yahoo.component.AbstractComponent;
+import com.yahoo.container.di.componentgraph.Provider;
+import com.yahoo.document.config.DocumentmanagerConfig;
+import com.yahoo.documentapi.DocumentAccess;
+import com.yahoo.documentapi.messagebus.MessageBusDocumentAccess;
+import com.yahoo.documentapi.messagebus.MessageBusParams;
+import com.yahoo.documentapi.messagebus.loadtypes.LoadTypeSet;
+import com.yahoo.vespa.config.content.LoadTypeConfig;
+
+/**
+ * Has a lazily populated reference to a {@link MessageBusDocumentAccess}.
+ *
+ * @author jonmv
+ */
+public class MessageBusDocumentAccessProvider extends AbstractComponent implements Provider<DocumentAccess> {
+
+    private final DocumentmanagerConfig documentmanagerConfig;
+    private final LoadTypeConfig loadTypeConfig;
+    private final Object monitor = new Object();
+    private boolean shutDown = false;
+    private DocumentAccess access = null;
+
+    @Inject
+    public MessageBusDocumentAccessProvider(DocumentmanagerConfig documentmanagerConfig, LoadTypeConfig loadTypeConfig) {
+        this.documentmanagerConfig = documentmanagerConfig;
+        this.loadTypeConfig = loadTypeConfig;
+    }
+
+    @Override
+    public DocumentAccess get() {
+        synchronized (monitor) {
+            if (access == null) {
+                access = new MessageBusDocumentAccess((MessageBusParams) new MessageBusParams(new LoadTypeSet(loadTypeConfig)).setDocumentmanagerConfig(documentmanagerConfig));
+                if (shutDown)
```

Comment:
Why creating it at all if shutdown is set. Can you just add it to the outer if as '&& !shutDown'

Reply:
Then you'd return `null`. I thought it better to preserve the existing error message you'd get when trying to use a shut down document access.

Reply:
Alternatively could throw something custom that was clear on this.

Reply:
Throw instead, no point in constructing mbus access and shut it down just to preserve error message in this case.
This should never happen, but throwing is probably a good idea.

Reply:
Can do. 'twas my first impulse as well. And, yes, this should never happen, unless someone breaks DI.